warehouse/manage/views/__init__.py

```diff
@@ -1495,14 +1518,24 @@ def constrain_environment(self):
                 workflow_filepath=publisher.workflow_filepath,
                 environment=form.constrained_environment_name.data,
             )
-
         else:
             self.request.session.flash(
                 "Can only constrain the environment for GitHub and GitLab publishers",
                 queue="error",
             )
             return self.default_response
 
+        # The user might have already manually created the new constrained publisher
+        # before clicking the magic link to constrain the existing publisher.
+        if self._check_publisher_exists_in_db(constrained_publisher):
```

Comment:
Feels like this can be `constrained_publisher.exists()` instead or something similar, so a) the logic is per-publisher and b) this is easy to re-use elsewhere.

Reply:
Done! See the last commit